app/models/invoice.rb

```diff
@@ -13,6 +13,8 @@ class Invoice < ApplicationRecord
   MAX_CHARGE_RETRIES = 3
   DECIMALS   = 2
   CHARGE_PRECISION   = 2
+  MIN_YEAR = 1980
+  MAX_YEAR = 2100
```

Comment:
I generally don't like to hardcode values. Also, choosing the limits is tricky, why these and not others? Wouldn't be better to make the limits dynamic? I mean, no before the the creation of the account, not after the present month.

Reply:
Agreed. Something like 
```ruby
validates :period, comparison: { less_than: ->() { Time.now }, greater_than: ->(invoice) { invoice.buyer_account.created_at }}
```

I'm inclined though to hardcode `greater_than` as it could at least theoretically cause N+1 queries. Or maybe add `if: ->(invoice) { invoice.will_save_change_to_period? }` to the validation to skip it when no changes will happen.

Reply:
Yeah, it's the question of "what makes sense"...

Previously we were not restricting years at all, so the provider could issue an invoice for a previous years, for example. Or the following years.
Probably these scenarios are useful, especially in December or January months :) Providers may need to rectify some invoices by issuing invoices for the period which is not current month.

Setting minimum to when the buyer account was created sounds reasonable, but not sure, maybe any existing invoices might break this rule - would need to check.

Reply:
The comparison validator only seems to be available in Rails 7 :smiling_face_with_tear: 

Also, `:period` is not a date, but rather a custom type, so I think the simplest would still be to use a custom validator.

Reply:
wrt custom validator - maybe `inclusion: { in: from_time..to_time }` can do the job?

wrt `period, that is strange. I see
```
  %I[due_on period issued_on last_charging_retry].each do |attr|
    attribute attr, :date
  end
```

Also in schema there is `t.date "period"`.

But then it is some active record type, not actually Time as you said and would not be possible to compare it to a range :/

Reply:
> Constants are part of public signature. Some may decide to use these values which I hope never happens for other purposes :)

How can that happen? may you explain, please?

Reply:
somebody decides to use `Invoice::MIN_YEAR` for example somewhere else in the codebase as some sort of verification. It may not be very likely but if we don't want it to happen, better not expose these values to others.

btw rubocop has too many constants error. I think it triggers on more than 3 or something.

Reply:
I understand, thanks.

Reply:
I actually don't think it's a bad thing to have such values (and this one in particular) exposed and used in some other places.

But I'll try to get rid of those.